tasmota/xsns_73_hp303b.ino

```diff
@@ -29,104 +29,113 @@
+  ret = HP303BSensor.measurePressureOnce(p, hp303b_cfg.oversampling);
   if (ret != 0)
     return false;
 
-  bhp303b_sensor.temperature = (float)ConvertTemp(t);
-  bhp303b_sensor.pressure = (float)ConvertPressure(p) / 100; //conversion to hPa
+  hp303b_sensor[hp303b_idx].temperature = (float)ConvertTemp(t);
+  hp303b_sensor[hp303b_idx].pressure = (float)ConvertPressure(p) / 100; //conversion to hPa
 
-  bhp303b_sensor.valid = SENSOR_MAX_MISS;
+  hp303b_sensor[hp303b_idx].valid = SENSOR_MAX_MISS;
   return true;
 }
 
 /********************************************************************************************/
 
 void HP303B_Detect(void)
 {
-  for (uint32_t i = 0; i < sizeof(bhp303b_addresses); i++)
+  for (uint32_t i = 0; i < HP303B_MAX_SENSORS; i++)
   {
-    if (I2cActive(bhp303b_addresses[i])) { return; }
-
-    bhp303b_sensor.address = bhp303b_addresses[i];
+    if (I2cActive(HP303B_START_ADDRESS + i)) { return; }
 
-    HP303BSensor.begin( bhp303b_sensor.address);
-
-    if (HP303B_Read())
+    if (HP303BSensor.begin(HP303B_START_ADDRESS + i))
     {
-      I2cSetActiveFound(bhp303b_sensor.address, bhp303b_sensor.name);
-      bhp303b_sensor.type = 1;
+      hp303b_sensor[hp303b_cfg.count].address = HP303B_START_ADDRESS + i;
+      I2cSetActiveFound(hp303b_sensor[hp303b_cfg.count].address, hp303b_cfg.types);
+      hp303b_cfg.count++;
       break;
     }
   }
 }
 
 void HP303B_EverySecond(void)
 {
-  if (uptime &1) {
-    if (!HP303B_Read()) {
-      AddLogMissed(bhp303b_sensor.name, bhp303b_sensor.valid);
+  for (uint32_t i = 0; i < hp303b_cfg.count; i++) {
+    if (uptime &1) {
+    if (!HP303B_Read(i)) {
+      AddLogMissed(hp303b_cfg.types, hp303b_sensor[i].valid);
     }
   }
+  }
 }
 
 void HP303B_Show(bool json)
 {
-  if (bhp303b_sensor.valid)
-  {
-    char str_temperature[33];
-    dtostrfd(bhp303b_sensor.temperature, Settings.flag2.temperature_resolution, str_temperature);
-    char str_pressure[33];
-    dtostrfd(bhp303b_sensor.pressure, Settings.flag2.pressure_resolution, str_pressure);
+  for (uint32_t i = 0; i < hp303b_cfg.count; i++) {
+    char sensor_name[12];
+    strlcpy(sensor_name, hp303b_cfg.types, sizeof(sensor_name));
+    if (hp303b_cfg.count > 1) {
+      snprintf_P(sensor_name, sizeof(sensor_name), PSTR("%s%c0x%02X"), sensor_name, IndexSeparator(), hp303b_sensor[i].address); // MCP9808-18, MCP9808-1A  etc.
+    }
 
-    if (json)
+    if (hp303b_sensor[i].valid)
     {
-      ResponseAppend_P(PSTR(",\"HP303B\":{\"" D_JSON_TEMPERATURE "\":%s,\"" D_JSON_PRESSURE "\":%s"), str_temperature,  str_pressure);
-      ResponseJsonEnd();
-#ifdef USE_DOMOTICZ
-      if (0 == tele_period)
+      char str_temperature[33];
+      dtostrfd(hp303b_sensor[i].temperature, Settings.flag2.temperature_resolution, str_temperature);
+      char str_pressure[33];
+      dtostrfd(hp303b_sensor[i].pressure, Settings.flag2.pressure_resolution, str_pressure);
+
+      if (json)
```

Comment:
quote from Theo:
domoticz and knx only support one temp sensor index you'l need to add a test for loop index 0 here. See latest change in bh1750.

Reply:
Check and comments added.

Reply:
ok, now we have tasmota support of 2 Sensors, but not yet in the libary.

you must change the libary to support reading functions with different I2C addresses.
at the moment, we have only support of one sensor with fix I2c reading functions in the libary.
i.e:
`HP303BSensor.measureTempOnce(t, hp303b_cfg.oversampling);`
(reads only from the address that was last initialized)

Now, we need a index for the reading functions:
`HP303BSensor.measureTempOnce(hp303b_sensor[hp303b_idx].address, t, hp303b_cfg.oversampling);`

the local variable for i2c address in the libary is `m_slaveAddress` 
i can help you to change the libary, if you want.

Reply:
I'm gonna make an attempt but I could use some advice.

Since I'm rewriting the library for Tasmota, do you think I should remove the continuous and interrupt code from the lib to reduce size?

The library also support SPI but I don't know how to test this, is this useful for Tasmota or should remove it?

Reply:
no, you don't must remove the code for SPI or other functions. The compiler compile only used functions, so there is no Advantage and you save no space. you don' must complete re-write the libary. 
We need only the 2 little changes and all should work.
But feel free to do it...

 you must change only the 2 functions:
`measureTempOnce()` and
`measurePressureOnce()`

add a Parameter for I2c-address in both functions and all is working with tasmota.
(note: each function of both is overloaded with 2 different Parameters)

Reply:
Ok that's what I'll do for now, but the startMeasure...() functions called by measure...Once() use the same global m_initFail and m_opMode variables. If we want to implement this correctly I think these should set per sensor as well.

Reply:
example for temp:
`int16_t LOLIN_HP303B::measureTempOnce(uint8_t slaveAddress, int32_t &result)
{
	return measureTempOnce(slaveAddress, result, m_tempOsr);
}`
and
`int16_t LOLIN_HP303B::measurePressureOnce(uint8_t slaveAddress; int32_t &result, uint8_t oversamplingRate)
{
  m_slaveAddress = slaveAddress;
	//start the measurement
.......`

should work.

don't forget to change the defines in header file...

Reply:
ah ok, thats bad with the `m_initFail and m_opMode`
variables.
I have a look tomorrow for it...
I hope, this is not a big Problem..

Reply:
or you make the implementation of more I2c addresses in a second Step.
You can change `#define HP303B_MAX_SENSORS    2 `
to 1 and do first a PR in Tasmota.

Reply:
ok, the `m_initFail` is not a Problem. you can delete the check of this variable from reading function, we have checked the init with the `begin()` function.
The `m_opMode` needs a little bit more investigation.

@arendst 
can you please check the driver as it is and you can merge it?

With the change `#define HP303B_MAX_SENSORS 1`, the driver should work with one sensor for now.
the support for two sensors comes later.
thanks in Advance.

Reply:
I committed the change to the lib. It would be strange if you connect 2 sensors and have them running in different operation mode. So this should not be an issue